lib/matrix/projection.nit

```diff
@@ -0,0 +1,153 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Services on `Matrix` to transform and project 3D coordinates
+module projection
+
+intrude import matrix
+
+redef class Matrix
+
+	# Create an orthogonal projection matrix
+	#
+	# `left, right, bottom, top, near, far` defines the world clip planes.
+	new orthogonal(left, right, bottom, top, near, far: Float)
+	do
+		var dx = right - left
+		var dy = top - bottom
+		var dz = far - near
+
+		assert dx != 0.0 and dy != 0.0 and dz != 0.0
+
+		var mat = new Matrix.identity(4)
+		mat[0, 0] = 2.0 / dx
+		mat[3, 0] = -(right + left) / dx
+		mat[1, 1] = 2.0 / dy
+		mat[3, 1] = -(top + bottom) / dy
+		mat[2, 2] = 2.0 / dz
+		mat[3, 2] = -(near + far) / dz
+		return mat
+	end
+
+	# Create a perspective transformation matrix
+	#
+	# Using the given vertical `field_of_view_y` in radians, the `aspect_ratio`
+	# and the `near`/`far` world distances.
+	new perspective(field_of_view_y, aspect_ratio, near, far: Float)
+	do
+		var frustum_height = (field_of_view_y/2.0).tan * near
+		var frustum_width = frustum_height * aspect_ratio
+
+		return new Matrix.frustum(-frustum_width, frustum_width,
+		                          -frustum_height, frustum_height,
+		                          near, far)
+	end
+
+	# Create a frustum transformation matrix
+	#
+	# `left, right, bottom, top, near, far` defines the world clip planes.
+	new frustum(left, right, bottom, top, near, far: Float)
```

Comment:
It's the matrix used to create the perspective effect in 3D games.

I prefer not linking to wikipedia in my doc, I thrust that the users know how to use Google!